src/lib/logstore/log_dev.cpp

```diff
@@ -435,32 +435,36 @@ bool LogDev::flush() {
         return false;
     }
 
-    LogGroup* lg = prepare_flush(new_idx - m_last_flush_idx + 4); // Estimate 4 more extra in case of parallel writes
-    if (sisl_unlikely(!lg)) {
-        THIS_LOGDEV_LOG(TRACE, "Log idx {} last_flush_idx {} prepare flush failed", new_idx, m_last_flush_idx);
-        return false;
-    }
-    auto sz = m_pending_flush_size.fetch_sub(lg->actual_data_size(), std::memory_order_relaxed);
-    HS_REL_ASSERT_GE((sz - lg->actual_data_size()), 0, "size {} lg size {}", sz, lg->actual_data_size());
-    off_t offset = m_vdev_jd->alloc_next_append_blk(lg->header()->total_size());
-    lg->m_log_dev_offset = offset;
-
-    HS_REL_ASSERT_NE(lg->m_log_dev_offset, INVALID_OFFSET, "log dev is full");
-    THIS_LOGDEV_LOG(TRACE, "Flushing log group data size={} at offset={} log_group={}", lg->actual_data_size(), offset,
-                    *lg);
-
-    HISTOGRAM_OBSERVE(logstore_service().m_metrics, logdev_flush_records_distribution, lg->nrecords());
-    HISTOGRAM_OBSERVE(logstore_service().m_metrics, logdev_flush_size_distribution, lg->actual_data_size());
+    for (; m_last_flush_idx < new_idx;) {
```

Comment:
I believe this change is driven by UT, could you share what is the case we need the `for loop` here and put it into the comment of code?

Reply:
the amount of logs which one logGroup can flush  has a upper limit.
pls see https://github.com/eBay/HomeStore/blob/314ee9ac42e7df2b4b0888554fe251aa42008b63/src/lib/logstore/log_group.cpp#L76


here we want to make sure all the log that need to be flushed will be definitely flushed , so we need this loop to create multiple log group if necessary

Reply:
cc @sanebay

Reply:
But even if we limit, it will be flushed in the next flush. Here if we flush in for loop, how is the ordering and flush completions works ? It becomes too complicated. Shouldnt you test make it work for this.

Reply:
only `logstore#on_write_completion` , which will be called after flushing a logGroup, will update the `m_tail_lsn`. 
when we truncate, we can not truncate exceeding `m_tail_lsn`.  

for example , we have a log store and we append upto lsn 100. now, if we flush only to 60, then  we can not truancate exceeding 60 , for example,  80. now , if we want to truncate to 80, we need to flush exceeding 80.

pls see
https://github.com/eBay/HomeStore/blob/3030851cebe24b99a070934bc67c5414a819a359/src/lib/logstore/log_store.cpp#L191

> how is the ordering and flush completions works

I do not get this question. when any logGroup is flushed, the flush completion will work for all the entry in the group.

Reply:
```
    /**
     * @brief Flush this log store (write/sync to disk) up to the sequence number
     *
     * @param seq_num Sequence number upto which logs are to be flushed. If not provided, will wait to flush all seq
     * numbers issued prior.
     */
    void flush(logstore_seq_num_t upto_seq_num = invalid_lsn());

```

I think the `HomeLogStore::flush`  definition state that we need to flush up to the given LSN, not best effort.  So we need to fix this somewhere.  Also the `LogDev::flush()` doesnt take any argument,  usually that means flush to the point when it was called, like the `std::flush()`

one step back, if we dont change `LogDev::flush()`,  we need to do like below in `HomeLogStore::flush` which is much less optimal 
```
 while (m_records.status(upto_lsn).is_active) {
     m_logdev->flush_under_guard();
     sleep(xxx);
}
```